aten/src/ATen/native/Blas.cpp

```diff
@@ -4,60 +4,35 @@
-    if (mat.stride(0) == 1 && lda_cond(mat.size(0), mat.size(1), mat.stride(1))) {
-      gemv<scalar_t>('n', mat.size(0), mat.size(1), alpha, mat.data_ptr<scalar_t>(), mat.stride(1),
-          vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
-    }
-    else if (mat.stride(1) == 1 && lda_cond(mat.size(1), mat.size(0), mat.stride(0))) {
-      gemv<scalar_t>('t', mat.size(1), mat.size(0), alpha, mat.data_ptr<scalar_t>(), mat.stride(0),
-          vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
-    }
-    else {
-      Tensor cmat = mat.contiguous();
-      gemv<scalar_t>('t', mat.size(1), mat.size(0), alpha, cmat.data_ptr<scalar_t>(), cmat.stride(0),
-          vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
-    }
-  });
-  return result;
-}
-
-Tensor &addmv_out(const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta, const Scalar& alpha, Tensor& result) {
-  { // scope of NoNamesGuard
-
-  at::NoNamesGuard guard;
-  result.resize_({mat.size(0)});
+namespace at {
+namespace meta {
+  TORCH_META_FUNC(addmv)(const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta, const Scalar& alpha) {
 
   Tensor self_ = self;
-  if (self.dim() == 0 || self.size(0) == 1) {
+  if (self.numel()==1 && self.dim()<=1) {
     self_ = self.expand({mat.size(0)});
   }
+  //if self was broadcastable, it's now correct size
+  //if it wasn't, it'll error out in the subsequent checks
 
   TORCH_CHECK((mat.dim() == 2 && vec.dim() == 1 && self_.dim() == 1),
-    "vector + matrix @ vector expected, got ", self_.dim(), ", ", mat.dim(), ", ", vec.dim());
-  TORCH_CHECK((mat.size(1) == vec.size(0) && mat.size(0) == self_.size(0)),
-    "size mismatch, get ", self_.size(0), ", ", mat.size(0), "x", mat.size(1), ",", vec.size(0));
+    "vector + matrix @ vector expected, got ", self.dim(), ", ", mat.dim(), ", ", vec.dim());
 
-  if (mat.numel() == 0) {
+  TORCH_CHECK((mat.size(1) == vec.size(0) && mat.size(0) == self_.size(0)),
+    "size mismatch, got ", self_.size(0), ", ", mat.size(0), "x", mat.size(1), ",", vec.size(0));
+
+  std::array<int64_t, 1> out_size{mat.sizes()[0]};
+  set_output(out_size, mat.options());
+  auto result = maybe_get_output();
+  //this check can fire for inplace op only, for all other versions result is guaranteed to be correct size
+  TORCH_CHECK(result.dim() == 1 && result.sizes()[0] == mat.sizes()[0], "output of addmv operation should be 1D with ",
+  "size equal to mat.size(0), yet got output size ", result.sizes(), " and mat.size(0) ", mat.size(0));
+  if (mat.numel() != 0){
+    if (!result.is_same(self_)) {
+      at::native::copy_(result, self_);
+    }
+  } else {
+    // shortcut for an empty matrix
     // By definition, when beta==0, values in self should be ignored. nans and infs
     // should not propagate
     if (beta.toComplexDouble() == 0.0) {
```

Comment:
I have no idea how at::cpu::mul_out below works for cuda inputs, but it does (and is preserved from current implementation). Still, of course neither `at::cpu::mul_out` nor `at::native::scalar_tensor` shouldn't be used in meta fn.